replica/table.cc

```diff
@@ -157,10 +157,21 @@ table::add_memtables_to_reader_list(std::vector<flat_mutation_reader_v2>& reader
         const tracing::trace_state_ptr& trace_state,
         streamed_mutation::forwarding fwd,
         mutation_reader::forwarding fwd_mr) const {
-    for (auto&& mt : *_compaction_group->memtables()) {
-        if (auto reader_opt = mt->make_flat_reader_opt(s, permit, range, slice, pc, trace_state, fwd, fwd_mr)) {
-            readers.emplace_back(std::move(*reader_opt));
+    auto add_memtables_from_cg = [&] (compaction_group& cg) mutable {
+        for (auto&& mt: *cg.memtables()) {
+            if (auto reader_opt = mt->make_flat_reader_opt(s, permit, range, slice, pc, trace_state, fwd, fwd_mr)) {
+                readers.emplace_back(std::move(*reader_opt));
+            }
         }
+    };
+    // point queries can optimized as they span a single compaction group.
+    if (range.is_singular() && range.start()->value().has_key()) {
+        const dht::ring_position& pos = range.start()->value();
+        add_memtables_from_cg(compaction_group_for_token(pos.token()));
+    }
```

Comment:
So the reserve() for the reader list is wrong in this case. We iterated on all compaction groups, but we should have picked just one.

Reply:
There's also a bug here. After adding memtables from a single group, I should have returned.